codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,266 @@
+        # 1487896,john-job-3157358
+        # 1478830,john-job-5234492
+        submitted_jobs = set()
+        jobs = self.run_command(['squeue', '-u', self.username, '--format', '%A,%j', '--noheader'])
+        for job in jobs.strip().split():
+            job_id, job_name = job.split(',')
+            if job_name.startswith(self.username) and self.args.job_name in job_name:
+                submitted_jobs.add(job_id)
+        return submitted_jobs
+
+    def get_worker_jobs(self):
+        """
+        Return a list of workers in RUNNING and PENDING state.
+        The current Slurm version on NLP cluster is 17.11.13-2, which doesn't have rest api provided.
+        """
+        # Documentation can be found at https://slurm.schedmd.com
+        # Since allocating resource for a worker may take a while, we periodically check
+        # for worker status and remove those workers that failed at starting phase.
+        jobs_to_remove = set()
+        for job_id in self.submitted_jobs:
+            job_acct = self.run_command(['sacct', '-j', job_id, '--format', 'state'])
+            if 'FAILED' in job_acct:
+                jobs_to_remove.add(job_id)
+                logger.error("Failed to start job {}".format(job_id))
+            elif 'COMPLETING' in job_acct or 'COMPLETED' in job_acct or 'CANCELLED' in job_acct:
+                jobs_to_remove.add(job_id)
+        self.submitted_jobs = self.submitted_jobs - jobs_to_remove
+        logger.info("Submitted jobs: ".format(self.submitted_jobs))
+
+        # Get all the Slurm workers that are submitted by SlurmWorkerManager and owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID:STATE (header won't be included with "--noheader" option)
+        # 1478828,PENDING
+        # 1478830,PENDING
+        jobs = self.run_command(['squeue', '-u', self.username, '--format', '%A,%T', '--noheader'])
+        jobs = jobs.strip().split()
+        logger.info(
+            'Workers: {}'.format(
+                ' '.join(job for job in jobs if job in self.submitted_jobs) or '(none)'
+            )
+        )
+
+        # Get all the RUNNING jobs that are owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID (header won't be included with "--noheader" option)
+        # 1478828
+        # 1478830
+        running_jobs = self.run_command(
+            ['squeue', '-u', self.username, '-t', 'RUNNING', '--format', '%A', '--noheader']
+        )
+        running_jobs = running_jobs.strip().split()
+
+        return [
+            WorkerJob(active=True) if job in running_jobs else WorkerJob(active=False)
+            for job in self.submitted_jobs
+        ]
+
+    def start_worker_job(self):
+        """
+        Start a CodaLab Slurm worker that submits batch job to Slurm
+        """
+        worker_id = uuid.uuid4().hex
+        # Set up worker directory
```

Comment:
I think this should be omitted—the worker directory isn't necessarily guaranteed to exist on the host that's running the slurm worker.

Reply:
Sorry, I mean the line with `work_dir.mkdir(parents=True, exist_ok=True)`.

Reply:
Hmm, what do you mean? It's currenlty coded to create any missing directories as necessary. See this [pathlib.Path.mkdir](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir)

Reply:
Consider the following (actual) example: I want to run workers on the jagupard machines, which each have their own separate filesystem. I want the codalab workers to use the directory `/scr/scr-with-most-space` on each of the jagupard machines. So, I set that as the input work dir to the slurmworkermanager. However, I'm launching the slurmworkermanager from sc. It sees that I want to use `/scr/scr-with-most-space` as the work dir, so it tries to make `/scr/scr-with-most-space` on sc, not on the worker node. I don't have permissions to do this operation on sc, so it fails.

In general, I think the temporary directories we use to store sbatch scripts and other artifacts of the slurmworkermanager should be divorced from the directory that the worker, when it lands on a slurm node, runs its jobs in.